core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryContextMapping.java

```diff
@@ -0,0 +1,260 @@
+     *     <li>String: <pre>&quot;string&quot;</pre></li>
+     *  </ul>
+     */
+    @Override
+    public Set<CharSequence> parseContext(ParseContext parseContext, XContentParser parser) throws IOException, ElasticsearchParseException {
+        final Set<CharSequence> contexts = new HashSet<>();
+        Token token = parser.currentToken();
+        if (token == Token.VALUE_STRING) {
+            contexts.add(parser.text());
+        } else if (token == Token.START_ARRAY) {
+            while ((token = parser.nextToken()) != Token.END_ARRAY) {
+                if (token == Token.VALUE_STRING) {
+                    contexts.add(parser.text());
+                } else {
+                    throw new ElasticsearchParseException("context array must have string values");
+                }
+            }
+        } else {
+            throw new ElasticsearchParseException("contexts must be a string or a list of strings");
+        }
+        return contexts;
+    }
+
+    @Override
+    public Set<CharSequence> parseContext(Document document) {
+        Set<CharSequence> values = null;
+        if (fieldName != null) {
+            IndexableField[] fields = document.getFields(fieldName);
+            values = new HashSet<>(fields.length);
+            for (IndexableField field : fields) {
+                values.add(field.stringValue());
+            }
+        }
+        return (values == null) ? Collections.<CharSequence>emptySet() : values;
+    }
+
+    /**
+     * Parse a list of {@link CategoryQueryContext}
+     * using <code>parser</code>. A QueryContexts accepts one of the following forms:
+     *
+     * <ul>
+     *     <li>Object: CategoryQueryContext</li>
+     *     <li>String: CategoryQueryContext value with prefix=false and boost=1</li>
+     *     <li>Array: <pre>[CategoryQueryContext, ..]</pre></li>
+     * </ul>
+     *
+     *  A CategoryQueryContext has one of the following forms:
+     *  <ul>
+     *     <li>Object: <pre>{&quot;context&quot;: <i>&lt;string&gt;</i>, &quot;boost&quot;: <i>&lt;int&gt;</i>, &quot;prefix&quot;: <i>&lt;boolean&gt;</i>}</pre></li>
+     *     <li>String: <pre>&quot;string&quot;</pre></li>
+     *  </ul>
+     */
+    @Override
+    public List<CategoryQueryContext> parseQueryContext(XContentParser parser) throws IOException, ElasticsearchParseException {
+        List<CategoryQueryContext> queryContexts = new ArrayList<>();
+        Token token = parser.nextToken();
+        if (token == Token.START_OBJECT || token == Token.VALUE_STRING) {
+            queryContexts.add(innerParseQueryContext(parser));
+        } else if (token == Token.START_ARRAY) {
+            while (parser.nextToken() != Token.END_ARRAY) {
+                queryContexts.add(innerParseQueryContext(parser));
+            }
+        }
+        return queryContexts;
+    }
+
+    private CategoryQueryContext innerParseQueryContext(XContentParser parser) throws IOException, ElasticsearchParseException {
```

Comment:
I wonder if you can use `ObjectParser` for this parsing? I can give you a hand if want with this?